core/store/src/trie/mem/updating.rs

```diff
@@ -43,6 +44,28 @@ pub enum UpdatedMemTrieNode {
     },
 }
 
+/// Keeps values and internal nodes accessed on updating memtrie.
+pub(crate) struct TrieAccesses {
+    /// Hashes and encoded trie nodes.
+    pub nodes: HashMap<CryptoHash, Arc<[u8]>>,
+    /// Hashes of accessed values - because values themselves are not
+    /// necessarily present in memtrie.
+    pub values: HashSet<CryptoHash>,
+}
```

Comment:
I don't fully understand what are the additional nodes that this change records.

AFAIU runtime updates the trie by calling `storage_write`, which always does a trie read before writing a value:
https://github.com/near/nearcore/blob/f2f6ed060625c92fa0581f475652999149d42ba5/runtime/near-vm-runner/src/logic/logic.rs#L2545

This trie read records all nodes that were accessed to reach this value, even in the case of memtries: https://github.com/near/nearcore/blob/f2f6ed060625c92fa0581f475652999149d42ba5/core/store/src/trie/mod.rs#L1296

Doesn't that record everything that is needed to prove execution of the contract? Why do we need an additional access log?

Reply:
Ah I guess it doesn't record nodes that are created when adding new values

Reply:
And this path is called only during contract execution. The gateway to update trie outside it is `set<T: BorshSerialize>(state_update: &mut TrieUpdate, key: TrieKey, value: &T)`.